### DISCUSSION TITLE
Debug information with types and sizes?

### DISCUSSION THREAD

**Discussion by clydeshaffer** [cc65/cc65#2317] [Q&A]
I have an emulator that I'm working on adding debug tools to. Currently I'm parsing out.map to get labelled values and addresses, but now I'm hitting the limit of this approach.

Namely, out.map doesn't include any size or type information, nor does it specify whether an entry is a function, variable, or constant.

It makes sense that this information would be out of the scope of out.map, so is there another way to generate this debug information at build time?

It'd also be handy if a code address could be associated with lines in a source file for setting breakpoints and such.

**Comment by pm100** [comment]
I have just written a debugger for sim65 (got fed up with firing up VICE) and was thinking along similar lines.

I posted two PR to start things rolling
- one expands macros in assembler listing
- second one outputs the assembler addresses with segment prefixes

These 2 things are needed to generated absolute loader data once the linker has finished, they are just a start.

BTW I would welcome feedback on the debugger, it is custom built for sim65. But could be adapted to other scenarios.
It has a few advanced features beyond basics
- watch for reads of memory not written to
- watch for read or write of specifc locations
- trap unbalanced stacks

its here https://github.com/pm100/db65

Missing
- write memory
- change registers
- all sorts of configuration
